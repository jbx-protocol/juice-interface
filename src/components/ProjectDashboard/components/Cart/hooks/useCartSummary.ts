import { useProjectCart } from 'components/ProjectDashboard/hooks'
import { NftRewardsContext } from 'contexts/NftRewards/NftRewardsContext'
import { useWallet } from 'hooks/Wallet'
import { useCallback, useContext, useMemo } from 'react'
import { formatCurrencyAmount } from 'utils/formatCurrencyAmount'

export const useCartSummary = () => {
  const { isConnected, connect } = useWallet()
  const cart = useProjectCart()
  const nftRewardsCtx = useContext(NftRewardsContext).nftRewards
  const tiers = useMemo(
    () => nftRewardsCtx.rewardTiers ?? [],
    [nftRewardsCtx.rewardTiers],
  )
  const nftRewards = useMemo(() => {
    return cart.nftRewards
      .map(nft => ({ ...tiers.find(tier => tier.id === nft.id), ...nft }))
      .filter(nft => nft !== undefined)
      .map(nft => ({
        fileUrl: nft.fileUrl,
        name: nft.name,
        quantity: nft.quantity,
        id: nft.id,
      }))
  }, [cart.nftRewards, tiers])

  const amountText = cart.totalAmount
    ? formatCurrencyAmount(cart.totalAmount)
    : undefined

  const resetCart = useCallback(() => {
    cart.dispatch({ type: 'reset' })
  }, [cart])

  const payProject = useCallback(() => {
    if (!isConnected) {
      connect()
      return
    }
    cart.dispatch({ type: 'openPayModal' })
  }, [cart, connect, isConnected])

  const showCurrencyOnCollapse = useMemo(() => {
    return cart.payAmount?.amount ? cart.payAmount.amount > 0 : false
  }, [cart.payAmount])

  return {
    amountText,
    currency: cart.totalAmount?.currency,
    nftRewards,
    walletConnected: isConnected,
    showCurrencyOnCollapse,
    resetCart,
    payProject,
  }
}
