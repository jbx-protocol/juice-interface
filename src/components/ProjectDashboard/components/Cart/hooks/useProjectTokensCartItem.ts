import {
  useProjectCart,
  useTokensPerEth,
} from 'components/ProjectDashboard/hooks'
import { useCallback } from 'react'

export const useProjectTokensCartItem = () => {
  const {
    totalAmount: payAmount,
    userIsReceivingTokens,
    dispatch,
  } = useProjectCart()
  const { receivedTickets } = useTokensPerEth(payAmount)

  const removeTokens = useCallback(
    () => dispatch({ type: 'removeTokens' }),
    [dispatch],
  )
  return {
    userIsReceivingTokens,
    receivedTickets,
    removeTokens,
  }
}
