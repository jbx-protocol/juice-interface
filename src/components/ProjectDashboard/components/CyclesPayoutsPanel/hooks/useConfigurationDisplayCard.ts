import { t } from '@lingui/macro'
import { useMemo } from 'react'

export const useConfigurationDisplayCard = (type: 'current' | 'upcoming') => {
  const title = useMemo(() => {
    if (type === 'current') {
      return t`Current cycle`
    }
    return t`Upcoming cycle`
  }, [type])

  return {
    title,
  }
}
