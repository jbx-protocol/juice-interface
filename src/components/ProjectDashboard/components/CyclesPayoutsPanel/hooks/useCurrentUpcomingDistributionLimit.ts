import {
  useProjectContext,
  useProjectMetadata,
} from 'components/ProjectDashboard/hooks'
import useProjectDistributionLimit from 'hooks/v2v3/contractReader/useProjectDistributionLimit'
import { useProjectUpcomingFundingCycle } from 'hooks/v2v3/contractReader/useProjectUpcomingFundingCycle'

export const useCurrentUpcomingDistributionLimit = (
  type: 'current' | 'upcoming',
) => {
  const { projectId } = useProjectMetadata()
  const {
    primaryETHTerminal,
    distributionLimit: currentDistributionLimit,
    distributionLimitCurrency: currentDistributionLimitCurrency,
  } = useProjectContext()

  const {
    data: upcomingFundingCycleData,
    loading: upcomingFundingCycleLoading,
  } = useProjectUpcomingFundingCycle({ projectId })
  const [upcomingFundingCycle] = upcomingFundingCycleData ?? []

  const {
    data: upcomingDistributionLimitData,
    loading: upcomingDistributionLimitLoading,
  } = useProjectDistributionLimit({
    projectId,
    configuration: upcomingFundingCycle?.configuration.toString(),
    terminal: primaryETHTerminal,
  })

  const [upcomingDistributionLimit, upcomingDistributionLimitCurrency] =
    upcomingDistributionLimitData ?? []

  if (type === 'current') {
    return {
      distributionLimit: currentDistributionLimit,
      distributionLimitCurrency: currentDistributionLimitCurrency,
      loading: false,
    }
  }
  return {
    distributionLimit: upcomingDistributionLimit,
    distributionLimitCurrency: upcomingDistributionLimitCurrency,
    loading: upcomingFundingCycleLoading || upcomingDistributionLimitLoading,
  }
}
