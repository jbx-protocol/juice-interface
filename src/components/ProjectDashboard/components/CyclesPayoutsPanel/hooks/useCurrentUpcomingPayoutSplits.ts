import {
  useProjectContext,
  useProjectMetadata,
} from 'components/ProjectDashboard/hooks'
import { ETH_PAYOUT_SPLIT_GROUP } from 'constants/splits'
import useProjectSplits from 'hooks/v2v3/contractReader/useProjectSplits'
import { useProjectUpcomingFundingCycle } from 'hooks/v2v3/contractReader/useProjectUpcomingFundingCycle'

export const useCurrentUpcomingPayoutSplits = (
  type: 'current' | 'upcoming',
) => {
  const { projectId } = useProjectMetadata()
  const { payoutSplits: currentPayoutSplits } = useProjectContext()

  const {
    data: upcomingFundingCycleData,
    loading: upcomingFundingCycleLoading,
  } = useProjectUpcomingFundingCycle({ projectId })
  const [upcomingFundingCycle] = upcomingFundingCycleData ?? []

  const { data: upcomingPayoutSplits, loading: upcomingProjectSplitsLoading } =
    useProjectSplits({
      projectId,
      splitGroup: ETH_PAYOUT_SPLIT_GROUP,
      domain: upcomingFundingCycle?.configuration?.toString(),
    })

  if (type === 'current') {
    return { splits: currentPayoutSplits, loading: false }
  }
  return {
    splits: upcomingPayoutSplits,
    loading: upcomingFundingCycleLoading || upcomingProjectSplitsLoading,
  }
}
