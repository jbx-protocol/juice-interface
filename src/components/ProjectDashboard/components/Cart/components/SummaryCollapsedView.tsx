import { TrashIcon } from '@heroicons/react/24/outline'
import { PlusIcon } from '@heroicons/react/24/solid'
import { Trans, t } from '@lingui/macro'
import { handleConfirmationDeletion } from 'components/ProjectDashboard/utils/modals'
import useMobile from 'hooks/useMobile'
import { useMemo } from 'react'
import useResizeObserver from 'use-resize-observer'
import { V2V3_CURRENCY_ETH } from 'utils/v2v3/currency'
import { SmallNftSquare } from '../../NftRewardsCard/SmallNftSquare'
import { CurrencyIcon } from '../../ui/CurrencyIcon'
import StackedComponents, {
  StackComponentItem,
} from '../../ui/StackedComponents'
import { useCartSummary } from '../hooks/useCartSummary'
import { SummaryPayButton } from './SummaryPayButton'

const STACK_COMPONENT_SIZE = '48px'
const STACK_COMPONENT_OFFSET = '20px'
const SUMMARY_GAP = '16px'

const CART_MAX_LIMIT = 10

export const SummaryCollapsedView = () => {
  const {
    amountText,
    currency,
    nftRewards,
    showCurrencyOnCollapse,
    resetCart,
  } = useCartSummary()
  const isMobile = useMobile()
  const { ref: containerRef, width: containerWidth = 1 } = useResizeObserver()
  const { ref: totalToPayRef, width: totalToPayWidth = 1 } = useResizeObserver()
  const { ref: internalSummaryRef, width: internalSummaryWidth = 1 } =
    useResizeObserver()

  const maxItems = useMemo(() => {
    const stackComponentSize = parseInt(STACK_COMPONENT_SIZE, 10)
    const stackComponentOffset = parseInt(STACK_COMPONENT_OFFSET, 10)
    const summaryGap = parseInt(SUMMARY_GAP, 10)
    const workableWidth =
      containerWidth -
      internalSummaryWidth -
      totalToPayWidth -
      summaryGap -
      (32 + 16) - // Chevron padding
      4

    const maxItemsBasedOnWidthAvailable = Math.max(
      Math.floor(workableWidth / (stackComponentSize - stackComponentOffset)),
      0,
    )

    const maxItems = Math.min(maxItemsBasedOnWidthAvailable, CART_MAX_LIMIT)

    return maxItems
  }, [containerWidth, internalSummaryWidth, totalToPayWidth])

  const CurrencyIconComponent = useMemo(() => {
    if (!showCurrencyOnCollapse) return null
    return {
      Component: CurrencyIcon,
      props: {
        className: 'h-full w-full border-4 border-white dark:border-slate-950',
        currency: currency ?? V2V3_CURRENCY_ETH,
      },
    }
  }, [currency, showCurrencyOnCollapse])

  const NftComponents = useMemo(
    () =>
      nftRewards.map(nft => ({
        Component: SmallNftSquare,
        props: {
          border: true,
          nftReward: nft,
          className: 'h-full w-full',
        },
      })),
    [nftRewards],
  )

  const StackComponents: StackComponentItem[] = useMemo(
    () =>
      (
        [CurrencyIconComponent, ...NftComponents].filter(
          Boolean,
        ) as StackComponentItem[]
      ).slice(0, maxItems),
    [CurrencyIconComponent, NftComponents, maxItems],
  )

  const omittedItemCount = useMemo(() => {
    const potentialCount = nftRewards.length + (showCurrencyOnCollapse ? 1 : 0)
    return Math.max(potentialCount - maxItems, 0)
  }, [maxItems, nftRewards.length, showCurrencyOnCollapse])

  return (
    <div
      className="flex w-full items-center justify-between gap-1 px-8 py-6"
      ref={containerRef}
    >
      <div
        data-testid="cart-summary-closed-view-summary"
        className="flex min-w-0 items-center gap-4"
      >
        <span
          className="font-heading text-xl font-medium md:text-2xl"
          ref={internalSummaryRef}
        >
          <Trans>Summary</Trans>
        </span>
        <StackedComponents
          components={StackComponents}
          size={STACK_COMPONENT_SIZE}
          offset={STACK_COMPONENT_OFFSET}
        />
        {omittedItemCount > 0 && (
          <span className="flex h-fit items-center justify-center gap-1 rounded-2xl bg-grey-100 py-0.5 pl-1.5 pr-2 text-xs font-medium text-grey-700 dark:bg-slate-600 dark:text-slate-200">
            <PlusIcon className="h-3 w-3 text-grey-500 dark:text-grey-300" />
            {omittedItemCount}
          </span>
        )}
      </div>
      {isMobile ? (
        <span ref={totalToPayRef} className="text-xl font-medium">
          {amountText}
        </span>
      ) : (
        <div
          ref={totalToPayRef}
          data-testid="cart-summary-closed-view-total"
          className="flex cursor-auto items-center gap-8"
          onClick={e => e.stopPropagation()}
        >
          <div className="flex items-center gap-4">
            <span className="whitespace-nowrap text-sm text-grey-500 dark:text-slate-200">
              <Trans>Total to pay</Trans>
            </span>
            <span className="text-2xl font-medium">{amountText}</span>
            <TrashIcon
              data-testid="cart-summary-closed-view-trash-icon"
              role="button"
              className="h-5 w-5 text-grey-400 dark:text-slate-300"
              onClick={handleConfirmationDeletion({
                type: t`payment`,
                description: t`Removing the payment will remove all the items from the cart. Are you sure you want to remove the payment?`,
                onConfirm: resetCart,
              })}
            />
          </div>
          <SummaryPayButton />
        </div>
      )}
    </div>
  )
}
