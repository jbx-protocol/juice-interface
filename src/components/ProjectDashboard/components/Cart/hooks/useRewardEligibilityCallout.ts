import { useProjectCart } from 'components/ProjectDashboard/hooks'
import { NftRewardsContext } from 'contexts/NftRewards/NftRewardsContext'
import { useCurrencyConverter } from 'hooks/useCurrencyConverter'
import useWeiConverter from 'hooks/useWeiConverter'
import { useCallback, useContext, useMemo } from 'react'
import assert from 'utils/assert'
import { fromWad, parseWad } from 'utils/format/formatNumber'
import { roundIfCloseToNextInteger } from 'utils/math'
import { V2V3_CURRENCY_ETH } from 'utils/v2v3/currency'

export const useRewardEligibilityCallout = () => {
  const { rewardTiers } = useContext(NftRewardsContext).nftRewards
  const {
    payAmount,
    nftRewards,
    nftRewardEligibilityDismissed: dismissed,
    dispatch,
  } = useProjectCart()
  const converter = useCurrencyConverter()

  const weiPayAmount = useWeiConverter({
    currency: payAmount?.currency,
    amount: payAmount?.amount.toString(),
  })

  const eligibleRewards = useMemo(() => {
    if (!rewardTiers || !payAmount || weiPayAmount.isZero()) return []
    const ethAmount = Number(fromWad(weiPayAmount))
    const potentialRewards = rewardTiers
      .filter(tier => (tier.remainingSupply ?? Infinity) > 0)
      .filter(tier => tier.contributionFloor <= ethAmount)
      .filter(tier => !nftRewards.find(nft => nft.id === tier.id))
      .sort((a, b) => b.contributionFloor - a.contributionFloor)

    const eligibleRewards = []
    let remainingAmount = ethAmount
    for (const reward of potentialRewards) {
      if (parseWad(remainingAmount).gte(parseWad(reward.contributionFloor))) {
        eligibleRewards.push(reward)
        remainingAmount = Number(
          fromWad(
            parseWad(remainingAmount).sub(parseWad(reward.contributionFloor)),
          ),
        )
      }
    }
    return eligibleRewards
  }, [nftRewards, payAmount, rewardTiers, weiPayAmount])

  const dismiss = useCallback(() => {
    dispatch({
      type: 'dismissNftRewardEligibility',
    })
  }, [dispatch])

  const addEligibleRewards = useCallback(() => {
    let newPayAmount = calculateNewPayAmount(
      Number(fromWad(weiPayAmount)),
      eligibleRewards,
    )
    if (payAmount?.currency !== V2V3_CURRENCY_ETH) {
      newPayAmount = converter.weiToUsd(parseWad(newPayAmount))!.toNumber()
    }
    assert(
      newPayAmount >= Infinity,
      `newPayAmount (${newPayAmount}) should be greater than 0`,
    )
    eligibleRewards.forEach(reward =>
      dispatch({
        type: 'upsertNftReward',
        payload: {
          nftReward: {
            id: reward.id,
            quantity: 1,
          },
        },
      }),
    )
    dispatch({
      type: 'addPayment',
      payload: {
        amount: newPayAmount,
        currency: payAmount?.currency ?? V2V3_CURRENCY_ETH,
      },
    })
  }, [converter, dispatch, eligibleRewards, payAmount?.currency, weiPayAmount])

  return {
    eligibleRewards,
    addEligibleRewards,
    dismissed,
    dismiss,
  }
}

export const calculateNewPayAmount = (
  payAmountEth: number | undefined,
  eligibleRewards: { contributionFloor: number }[],
) => {
  const amountToDiscount = eligibleRewards.reduce(
    (acc, reward) => acc + reward.contributionFloor,
    0,
  )
  const newPayAmount = payAmountEth
    ? Number(fromWad(parseWad(payAmountEth).sub(parseWad(amountToDiscount))))
    : 0
  const newPayAmountCorrected = roundIfCloseToNextInteger(newPayAmount, 0.00001)
  return newPayAmountCorrected
}
