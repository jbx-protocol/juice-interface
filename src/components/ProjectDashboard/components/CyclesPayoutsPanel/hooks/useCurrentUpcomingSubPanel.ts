import { t } from '@lingui/macro'
import {
  useProjectContext,
  useProjectMetadata,
} from 'components/ProjectDashboard/hooks'
import { useFundingCycleCountdown } from 'components/ProjectDashboard/hooks/useFundingCycleCountdown'
import { timeSecondsToDateString } from 'components/ProjectDashboard/utils/timeSecondsToDateString'
import { useProjectUpcomingFundingCycle } from 'hooks/v2v3/contractReader/useProjectUpcomingFundingCycle'
import { useMemo } from 'react'

export const useCurrentUpcomingSubPanel = (type: 'current' | 'upcoming') => {
  const { projectId } = useProjectMetadata()
  const {
    fundingCycle,
    loading: { fundingCycleLoading },
  } = useProjectContext()
  const {
    data: upcomingFundingCycleData,
    loading: upcomingFundingCycleLoading,
  } = useProjectUpcomingFundingCycle({ projectId })
  const [upcomingFundingCycle] = upcomingFundingCycleData ?? []
  const { timeRemainingText } = useFundingCycleCountdown()

  const cycleNumber = useMemo(() => {
    if (type === 'current') {
      return fundingCycle?.number.toNumber()
    }
    return fundingCycle?.number ? fundingCycle.number.toNumber() + 1 : undefined
  }, [fundingCycle?.number, type])

  const cycleUnlocked = useMemo(() => {
    if (type === 'current') {
      return fundingCycle?.duration?.isZero() ?? true
    }
    return upcomingFundingCycle?.duration?.isZero() ?? true
  }, [fundingCycle?.duration, type, upcomingFundingCycle?.duration])

  const upcomingCycleLength = useMemo(() => {
    if (!upcomingFundingCycle) return
    if (cycleUnlocked) return '-'
    return timeSecondsToDateString(
      upcomingFundingCycle.duration.toNumber(),
      'short',
    )
  }, [cycleUnlocked, upcomingFundingCycle])

  const status = useMemo(() => {
    if (cycleUnlocked) return t`Unlocked`
    return t`Locked`
  }, [cycleUnlocked])

  const remainingTime = useMemo(() => {
    if (cycleUnlocked) return '-'
    return timeRemainingText
  }, [cycleUnlocked, timeRemainingText])

  // Short circuit current for faster loading
  if (type === 'current') {
    if (fundingCycleLoading) return { loading: true, type }
    return {
      loading: false,
      type,
      cycleNumber,
      status,
      remainingTime,
    }
  }

  if (fundingCycleLoading || upcomingFundingCycleLoading)
    return {
      loading: true,
      type,
    }

  return {
    loading: false,
    type,
    cycleNumber,
    status,
    cycleLength: upcomingCycleLength,
  }
}
