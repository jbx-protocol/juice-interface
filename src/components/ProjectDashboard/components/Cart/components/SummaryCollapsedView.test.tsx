/**
 * @jest-environment jsdom
 */

import { fireEvent, render, waitFor } from '@testing-library/react'
import { V2V3_CURRENCY_ETH } from 'utils/v2v3/currency'
import { useCartSummary } from '../hooks/useCartSummary'
import { SummaryCollapsedView } from './SummaryCollapsedView'

jest.mock('use-resize-observer', () => ({
  __esModule: true,
  default: jest.fn(() => ({
    ref: jest.fn(),
    width: 100,
    height: 100,
  })),
}))
jest.mock('../hooks/useCartSummary')

describe('SummaryCollapsedView', () => {
  const DefaultUseCartSummary = {
    amountText: '0.0000 ETH',
    currency: V2V3_CURRENCY_ETH,
    nftRewards: [],
    resetCart: jest.fn(),
    payProject: jest.fn(),
  }
  beforeEach(() => {
    ;(useCartSummary as jest.Mock).mockReturnValue(DefaultUseCartSummary)
    DefaultUseCartSummary.resetCart.mockClear()
    DefaultUseCartSummary.payProject.mockClear()
  })

  it('should render correctly', () => {
    const { container } = render(<SummaryCollapsedView />)
    expect(container).toMatchSnapshot()
  })

  it('should call removePay when trash icon is clicked', () => {
    const { getByTestId, getByRole } = render(<SummaryCollapsedView />)
    const trashIcon = getByTestId('cart-summary-closed-view-trash-icon')
    fireEvent.click(trashIcon)
    waitFor(() => {
      const modalButton = getByRole('button')
      fireEvent.click(modalButton)
      expect(DefaultUseCartSummary.resetCart).toHaveBeenCalled()
    })
  })
})
