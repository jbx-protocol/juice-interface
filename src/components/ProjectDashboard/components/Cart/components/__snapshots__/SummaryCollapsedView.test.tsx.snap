// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SummaryCollapsedView should render correctly 1`] = `
<div>
  <div
    class="flex w-full items-center justify-between gap-1 px-8 py-6"
  >
    <div
      class="flex min-w-0 items-center gap-4"
      data-testid="cart-summary-closed-view-summary"
    >
      <span
        class="font-heading text-xl font-medium md:text-2xl"
      >
        Summary
      </span>
      <div
        class="relative z-0 flex min-w-0 items-center"
      />
    </div>
    <div
      class="flex cursor-auto items-center gap-8"
      data-testid="cart-summary-closed-view-total"
    >
      <div
        class="flex items-center gap-4"
      >
        <span
          class="whitespace-nowrap text-sm text-grey-500 dark:text-slate-200"
        >
          Total to pay
        </span>
        <span
          class="text-2xl font-medium"
        >
          0.0000 ETH
        </span>
        <svg
          aria-hidden="true"
          class="h-5 w-5 text-grey-400 dark:text-slate-300"
          data-testid="cart-summary-closed-view-trash-icon"
          fill="none"
          role="button"
          stroke="currentColor"
          stroke-width="1.5"
          viewBox="0 0 24 24"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M14.74 9l-.346 9m-4.788 0L9.26 9m9.968-3.21c.342.052.682.107 1.022.166m-1.022-.165L18.16 19.673a2.25 2.25 0 01-2.244 2.077H8.084a2.25 2.25 0 01-2.244-2.077L4.772 5.79m14.456 0a48.108 48.108 0 00-3.478-.397m-12 .562c.34-.059.68-.114 1.022-.165m0 0a48.11 48.11 0 013.478-.397m7.5 0v-.916c0-1.18-.91-2.164-2.09-2.201a51.964 51.964 0 00-3.32 0c-1.18.037-2.09 1.022-2.09 2.201v.916m7.5 0a48.667 48.667 0 00-7.5 0"
            stroke-linecap="round"
            stroke-linejoin="round"
          />
        </svg>
      </div>
      <button
        class="ant-btn ant-btn-primary"
        type="button"
      >
        <span>
          Connect wallet
        </span>
      </button>
    </div>
  </div>
</div>
`;
